### PR TITLE
[Service Bus] Avoid rethrowing exception with throw ex syntax when exception unchanged

### PR DESCRIPTION
**Pull request by danielmarbach**
Attempt to address #11639

image

## Before

```
    [TestFixture]
    public class AmqpExceptionHelperTests
    {
        [Test]
        public void Translates()
        {
            try
            {
                MethodThatThrows();
            }
            catch (Exception e)
            {
                throw AmqpExceptionHelper.TranslateException(e);
            }
        }

        private static void MethodThatThrows()
        {
            throw new InvalidOperationException();
        }
     }
```

leads to 

```
Azure.Messaging.ServiceBus.Tests.Amqp.AmqpExceptionHelperTests.Translates

System.AggregateException : One or more errors occurred. (Rethrown)
  ----> NUnit.Framework.Internal.NUnitException : Rethrown
  ----> System.InvalidOperationException : Operation is not valid due to the current state of the object.
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at NUnit.Framework.Internal.Commands.TimeoutCommand.Execute(TestExecutionContext context) in D:\a\1\s\src\NUnitFramework\framework\Internal\Commands\TimeoutCommand.cs:line 102
--NUnitException
   at NUnit.Framework.Internal.Reflect.InvokeMethod(MethodInfo method, Object fixture, Object[] args) in D:\a\1\s\src\NUnitFramework\framework\Internal\Reflect.cs:line 274
   at NUnit.Framework.Internal.Commands.TestMethodCommand.Execute(TestExecutionContext context) in D:\a\1\s\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs:line 64
   at NUnit.Framework.Internal.Commands.TimeoutCommand.<>c__DisplayClass2_0.<Execute>b__0() in D:\a\1\s\src\NUnitFramework\framework\Internal\Commands\TimeoutCommand.cs:line 102
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--InvalidOperationException
   at Azure.Messaging.ServiceBus.Tests.Amqp.AmqpExceptionHelperTests.Translates() in C:\p\azure-sdk-for-net\sdk\servicebus\Azure.Messaging.ServiceBus\tests\Amqp\AmqpExceptionHelperTests.cs:line 22


````

## After

```
    [TestFixture]
    public class AmqpExceptionHelperTests
    {
        [Test]
        public void Translates()
        {
            try
            {
                MethodThatThrows();
            }
            catch (Exception e)
            {
                ExceptionDispatchInfo.Capture(AmqpExceptionHelper.TranslateException(e))
                    .Throw();
                throw;
            }
        }

        private static void MethodThatThrows()
        {
            throw new InvalidOperationException();
        }
    }
```

leads to

```
Azure.Messaging.ServiceBus.Tests.Amqp.AmqpExceptionHelperTests.Translates

System.AggregateException : One or more errors occurred. (Rethrown)
  ----> NUnit.Framework.Internal.NUnitException : Rethrown
  ----> System.InvalidOperationException : Operation is not valid due to the current state of the object.
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at NUnit.Framework.Internal.Commands.TimeoutCommand.Execute(TestExecutionContext context) in D:\a\1\s\src\NUnitFramework\framework\Internal\Commands\TimeoutCommand.cs:line 102
--NUnitException
   at NUnit.Framework.Internal.Reflect.InvokeMethod(MethodInfo method, Object fixture, Object[] args) in D:\a\1\s\src\NUnitFramework\framework\Internal\Reflect.cs:line 274
   at NUnit.Framework.Internal.Commands.TestMethodCommand.Execute(TestExecutionContext context) in D:\a\1\s\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs:line 64
   at NUnit.Framework.Internal.Commands.TimeoutCommand.<>c__DisplayClass2_0.<Execute>b__0() in D:\a\1\s\src\NUnitFramework\framework\Internal\Commands\TimeoutCommand.cs:line 102
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--InvalidOperationException
   at Azure.Messaging.ServiceBus.Tests.Amqp.AmqpExceptionHelperTests.MethodThatThrows() in C:\p\azure-sdk-for-net\sdk\servicebus\Azure.Messaging.ServiceBus\tests\Amqp\AmqpExceptionHelperTests.cs:line 31
   at Azure.Messaging.ServiceBus.Tests.Amqp.AmqpExceptionHelperTests.Translates() in C:\p\azure-sdk-for-net\sdk\servicebus\Azure.Messaging.ServiceBus\tests\Amqp\AmqpExceptionHelperTests.cs:line 19
--- End of stack trace from previous location where exception was thrown ---
   at Azure.Messaging.ServiceBus.Tests.Amqp.AmqpExceptionHelperTests.Translates() in C:\p\azure-sdk-for-net\sdk\servicebus\Azure.Messaging.ServiceBus\tests\Amqp\AmqpExceptionHelperTests.cs:line 23
```

The ugly part is the non-reachable throw statements to make the compiler happy

### CODE REVIEW

**Comment by danielmarbach** [comment]
Because the `AmqpExceptionHelper` was not yet under tests I decided to not add tests given that I kept the control flow exactly the same.

**Comment by JoshLove-msft** [comment]
> The ugly part is the non-reachable throw statements to make the compiler happy

So what is the compiler complaining about when you don't include the throw statement?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs:143]: nit: we have a VS warning to use explicit types except when using literals or constructors.

**Review by JoshLove-msft — COMMENTED** [review]
What are your thoughts on inverting the pattern to be:
```
                ExceptionDispatchInfo.Capture(AmqpExceptionHelper.TranslateException(e))
                    .Throw();
```
The benefit is that we wouldn't have to propagate the ExceptionDispatchInfo into AmqpExceptionHelper, assuming that this still works as expected.

**Comment by danielmarbach** [comment]
> So what is the compiler complaining about when you don't include the throw statement?

Return statement missing

**Comment by danielmarbach** [comment]
> 
> 
> What are your thoughts on inverting the pattern to be:
> 
> ```
>                 ExceptionDispatchInfo.Capture(AmqpExceptionHelper.TranslateException(e))
>                     .Throw();
> ```
> 
> The benefit is that we wouldn't have to propagate the ExceptionDispatchInfo into AmqpExceptionHelper, assuming that this still works as expected.

Chnanged the approach to what you suggested because it is much leaner

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs:143]: I'd suggest sliding the `ExceptionDispatchInfo.Capture` to line 158 when you're ready to throw.  The code on lines 144-157 is concerned only with the exception, not the captured context.   Since the `throw` only takes place when not retrying, there's no benefit to having the capture unless you know that you're not retrying.

**Review by jsquire — APPROVED** [review]
LGTM.   Thank you for your contribution, @danielmarbach.   I'm going to defer to @JoshLove-msft and @ShivangiReja on the final approval and merge decision.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs
Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusRetryPolicy.cs:143]: Good suggestion. Will do

**Comment by JoshLove-msft** [comment]
@danielmarbach any chance you can do a rebase? A new yml file was added to workaround a live test run issue that is not getting picked up when running the live tests against this PR.

**Comment by JoshLove-msft** [comment]
> @danielmarbach any chance you can do a rebase? A new yml file was added to workaround a live test run issue that is not getting picked up when running the live tests against this PR.

Actually, hold off on this.. There was a bug with the path being used.

**Pull request merged**
